src/Stratis.Bitcoin/P2P/Peer/NetworkPeerFactory.cs

```diff
@@ -32,16 +33,18 @@ public interface INetworkPeerFactory
         /// <param name="myVersion">Version of the protocol that the node supports.</param>
         /// <param name="isRelay">Whether the remote peer should announce relayed transactions or not. See <see cref="VersionPayload.Relay"/> for more information.</param>
         /// <param name="cancellation">Cancallation token that allows to interrupt establishing of the connection.</param>
+        /// <param name="connectedPeersHelper">Maintains a list of connected peers and ensures their proper disposal. Or <c>null</c> if case disposal should be handled from user code.</param>
         /// <returns>Network peer connected to the specified counterparty.</returns>
-        Task<INetworkPeer> CreateConnectedNetworkPeerAsync(string endPoint, ProtocolVersion myVersion = ProtocolVersion.PROTOCOL_VERSION, bool isRelay = true, CancellationToken cancellation = default(CancellationToken));
+        Task<INetworkPeer> CreateConnectedNetworkPeerAsync(string endPoint, ProtocolVersion myVersion = ProtocolVersion.PROTOCOL_VERSION, bool isRelay = true, CancellationToken cancellation = default(CancellationToken), ConnectedPeersHelper connectedPeersHelper = null);
 
         /// <summary>
         /// Creates a new network peer which is connected to a specified counterparty.
         /// </summary>
         /// <param name="peerEndPoint">Address and port of the counterparty to connect to.</param>
         /// <param name="parameters">Parameters specifying how the connection with the counterparty should be established, or <c>null</c> to use default parameters.</param>
+        /// <param name="connectedPeersHelper">Maintains a list of connected peers and ensures their proper disposal. Or <c>null</c> if case disposal should be handled from user code.</param>
         /// <returns>Network peer connected to the specified counterparty.</returns>
-        Task<INetworkPeer> CreateConnectedNetworkPeerAsync(IPEndPoint peerEndPoint, NetworkPeerConnectionParameters parameters = null);
+        Task<INetworkPeer> CreateConnectedNetworkPeerAsync(IPEndPoint peerEndPoint, NetworkPeerConnectionParameters parameters = null, ConnectedPeersHelper connectedPeersHelper = null);
```

Comment:
`ConnectedPeersHelper` is a singleton that lives through out the node lifetime so you could just inject it to the factory.

Reply:
That was the first implementation of the ConnectedPeersHelper (add as a singleton, inject to the factory and add a bool `disposeOnDisconnect` that will make the factory use the helper, dispose the helper on ConnectionManager disposal).

Don't remember why we didn't stick with that solution, but for sure I don't mind it being that way.

Reply:
It is not a Singleton. Multiple instances exist. As explained above this class is used when a component wants to create network peers and does not want to manage the disposal and their tracking. So each such component will have its helper for this tracking and disposing.